transformer_engine/pytorch/attention.py

```diff
@@ -1112,8 +1179,63 @@ def forward(
             attn_mask_type = self.attn_mask_type
 
         assert (key_layer.shape[-2] == self.num_gqa_groups_per_partition
-                and value_layer.shape[-2] == self.num_gqa_groups_per_partition
-                ), f"Keys and values must have {self.num_gqa_groups} heads!"
+            and value_layer.shape[-2] == self.num_gqa_groups_per_partition
+            ), f"Keys and values must have num_gqa_group = {self.num_gqa_groups} heads!"
+        assert (qkv_format in ['sbhd', 'bshd', 'thd']
+            ), "DotProductAttention only supports qkv_format = {'sbhd', 'bshd', 'thd'}!"
+
+        if qkv_format == 'thd':
+            assert (all(len(x.shape) == 3 for x in (query_layer, key_layer, value_layer))
+                ), "Queries, keys and values must be 3D tensors when qkv_format = thd!"
+            assert (cu_seqlens_q is not None and cu_seqlens_kv is not None
+                ), "cu_seqlens_q and cu_seqlens_kv can not be None when qkv_format = thd!"
+            assert (cu_seqlens_q.shape == cu_seqlens_kv.shape
+                and len(cu_seqlens_q.shape) == 1
+                and len(cu_seqlens_kv.shape) == 1
+                ), "cu_seqlens_q and cu_seqlens_q must both have shape [batch_size + 1]!"
+            assert (cu_seqlens_q.dtype == torch.int32
+                and cu_seqlens_kv.dtype == torch.int32
+                ), "cu_seqlens_q and cu_seqlens_q must both be in dtype torch.int32!"
+            batch_size = len(cu_seqlens_q) - 1
+            seqlens_q = cu_seqlens_q[1:] - cu_seqlens_q[:-1]
+            seqlens_kv = cu_seqlens_kv[1:] - cu_seqlens_kv[:-1]
+            max_seqlen_q = seqlens_q.max().item()
+            max_seqlen_kv = seqlens_kv.max().item()
+            if (all(seqlens_q == seqlens_q[0])
+                and all(seqlens_kv == seqlens_kv[0])
+                and query_layer.shape[0] == batch_size * seqlens_q[0]
+                and key_layer.shape[0] == batch_size * seqlens_kv[0]):
```

Comment:
This check involves a GPU sync and it doesn't affect correctness, only performance. Perhaps it's not worth protecting the users from themselves in this case.

Reply:
Sure, I'll remove the logic for now. The original concern was that users might historically be used to the (bxs, 3, h, d) format and in the new design we want to distinguish that from (t, 3, h, d) where t3hd should be only used when s is not the same in a batch. But I think performance is a valid concern here so I'll remove it for now. Thanks!